src/server/frontend_xwayland/xwayland_server.cpp

```diff
@@ -86,52 +98,109 @@ mf::XWaylandServer::~XWaylandServer()
     close(socket_fd);
 }
 
+namespace
+{
+auto wl_client_for_xwayland(std::shared_ptr<mf::WaylandConnector> const& wlc, int wl_client_server_fd)
+-> wl_client*
+{
+    using namespace std::chrono_literals;
+
+    wl_client* client = nullptr;
+    std::mutex client_mutex;
+    std::condition_variable client_ready;
+
+    wlc->run_on_wayland_display([wl_client_server_fd, &client, &client_mutex, &client_ready](wl_display* display)
+        {
+            std::lock_guard<std::mutex> lock{client_mutex};
+            client = wl_client_create(display, wl_client_server_fd);
+            client_ready.notify_all();
+        });
+
+    std::unique_lock<std::mutex> lock{client_mutex};
+    if (!client_ready.wait_for(lock, 10s, [&]{ return client; }))
+    {
+        // "Shouldn't happen" but this is better than hanging.
+        mir::fatal_error("Failed to create wl_client for Xwayland");
+    }
+    return client;
+}
+
+bool spin_wait_for(sig_atomic_t& xserver_ready)
+{
+    using namespace std::chrono_literals;
+    auto const end = std::chrono::steady_clock::now() + 5s;
+
+    while (std::chrono::steady_clock::now() < end && !xserver_ready)
+    {
+        std::this_thread::sleep_for(100ms);
+    }
+
+    return xserver_ready;
+}
+}
+
 void mf::XWaylandServer::spawn()
 {
+    set_thread_name("XWaylandServer::spawn");
+
     enum { server, client, size };
     int wl_client_fd[size], wm_fd[size];
 
+retry_all:
```

Comment:
Alan: mumble mumble bad C code
Also Alan: *introduces gotos into an already complex function*

Surely there's a better way?

Reply:
You don't agree this is less bad than the recursive calls to `spawn()` that were scattered around? (And didn't work reliably.)

At least the funky logic is all in one place and most of the routine logic separated out. (Thanks for starting that off.)

Reply:
Quite possibly a net improvement, but introduces too much new complexity for me to comfortably call all issues pre-existing. Would it be a burden to abstract a single "attempt" into a function and call that in a loop?

Reply:
How about I split out https://github.com/MirServer/mir/pull/1152/commits/78a9f8f340a18af33ab2ff973606ccf8f051bcd9 and we land that separately?

Reply:
Sounds good